### PR TITLE
@/charterafrica/pages/ErrorPage #319

### LINKED ISSUE #319: @/charterafrica/pages/ErrorPage
## Description

Implement the error pages per [designs](https://www.figma.com/file/gbnaO1FYPq16V2gAFbLOkl/CfA---EPD-Charter-Africa---Sep-2022?node-id=9313%3A29587&t=gHql57tAK6HPaLGe-0). Things to keep in mind:

  1. Content of these pages should be editable in the CMS to allow localisation.
  2. 404 [needs to be static](https://nextjs.org/docs/advanced-features/custom-error-page#404-page). Current config seem to have issue with static building (could be because we run Payload and Next.js in the same server) & hence we may need a temporary solution e.g. hard-code 404 content to the page.
  3. BackgrondBox component can be reused here by customising `bgcolor` and `backgroundImage`.

### PR DESCRIPTION
**Pull request by koechkevin**
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #319 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Screenshots
image

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: apps/charterafrica/src/components/ErrorPage/ErrorPage.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    title = "",
+    createdAt,
+    updatedAt,
+    action,
+    ...other
+  } = props;
+
+  const link =
+    action &&
+    mapLinkTypeToHref({
```
Comment (kilemensi) [inline apps/charterafrica/src/components/ErrorPage/ErrorPage.js:26]: Can't think of a single reason why we would need this **_on_** front-end.

**Review by kilemensi — COMMENTED** [review]
👍🏽 

-----

Quick review; I see we're using `getStaticProps` in error pages but as I mentioned in the [issue](https://github.com/CodeForAfrica/ui/issues/319), this doesn't work. Have you tried running `pnpm build`?

File: apps/charterafrica/src/components/ErrorPage/ErrorPage.js (hunk 1 above)
Comment (koechkevin) [reply apps/charterafrica/src/components/ErrorPage/ErrorPage.js:26]: Could we be forced to use REST API instead of payload import?🤔.

Because footer and header needs data from payload.

Comment (kilemensi) [reply apps/charterafrica/src/components/ErrorPage/ErrorPage.js:26]: We may have to but I'm hoping there is a fix for our current situation @koechkevin ... local APIs are just too nice.

**Comment by kilemensi** [comment]
My two cents on the approach @koechkevin: Not sure making Errors a collection is a good idea or at least I don't see any benefit.

Error pages are just pages like any other page. Error details can be a Block that is added to a page. This way, 404 on front-end would then just fetch a page `slug=404` from CMS and that will be that.

**Comment by kilemensi** [comment]
Making progress @koechkevin ?

File: apps/charterafrica/.env.template [hunk 2]
```diff
@@ -20,6 +20,8 @@ NEXT_PUBLIC_APP_URL
 NEXT_PUBLIC_DEFAULT_LOCALE
 NEXT_PUBLIC_IMAGE_DOMAINS
 NEXT_PUBLIC_IMAGE_UNOPTIMIZED
+NEXT_PUBLIC_APP_AUTH_EMAIL
+NEXT_PUBLIC_APP_AUTH_PASSWORD
```
Comment (kilemensi) [inline apps/charterafrica/.env.template:24]: Do we need any auth? I believe all collections allow anyone to read e.g. for Pages: https://github.com/CodeForAfrica/ui/blob/1462db9951eb2fd1dc164dcc83753a212b21facc/apps/charterafrica/src/payload/collections/Pages.js#L19-L21

File: apps/charterafrica/package.json [hunk 3]
```diff
@@ -61,6 +61,7 @@
     "nodemon": "^2.0.20",
     "payload": "^1.6.4",
     "prop-types": "^15.8.1",
+    "qs": "^6.11.0",
```
Comment (kilemensi) [inline apps/charterafrica/package.json:64]: Do we need more than what [URLSearchParams](https://nodejs.org/docs/latest-v16.x/api/url.html#class-urlsearchparams) provides?

File: apps/charterafrica/src/components/ErrorPage/ErrorPage.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    title = "",
+    createdAt,
+    updatedAt,
+    blockType,
+    action,
+    ...other
+  } = props;
+  return (
+    <BackgroundBox
+      {...other}
```
Comment (kilemensi) [inline apps/charterafrica/src/components/ErrorPage/ErrorPage.js:25]:
```suggestion
```

Avoid this. For example, if you've received props with other params like id, etc., they'll all end up added to DOM.

File: apps/charterafrica/src/components/ErrorPage/ErrorPage.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import { Link } from "@commons-ui/next";
+import { Button } from "@mui/material";
+import PropTypes from "prop-types";
+import React from "react";
+
+import { secondary, neutral } from "@/charterafrica/colors";
+import BackgroundBox from "@/charterafrica/components/BackgroundBox";
+import RichText from "@/charterafrica/components/RichText";
+
+const ErrorPage = React.forwardRef(function Error(props, ref) {
```
Comment (kilemensi) [inline apps/charterafrica/src/components/ErrorPage/ErrorPage.js:12]: This doesn't look like a Page to be... It should probably just be named `Error` i.e. just another component.

File: apps/charterafrica/src/components/ErrorPage/ErrorPage.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        height: "100%",
+        width: "100%",
+        position: { md: "absolute", xs: "relative" },
+        backgroundSize: "50% 100%",
+        ...props?.sx,
+      }}
+      ref={ref}
+    >
+      <Section sx={{ p: 10 }}>
+        <RichTypography color={neutral[900]} textAlign="center" variant="h1">
```
Comment (kilemensi) [inline apps/charterafrica/src/components/ErrorPage/ErrorPage.js:42]:
```suggestion
        <RichTypography color="neutral.dark" textAlign="center" variant="h1">
```

File: apps/charterafrica/src/components/ErrorPage/ErrorPage.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+          {statusCode}
+        </RichTypography>
+        <RichTypography
+          sx={{ mt: 2.5, textTransform: "uppercase" }}
```
Comment (kilemensi) [inline apps/charterafrica/src/components/ErrorPage/ErrorPage.js:46]:
```suggestion
          mt={2.5}
          textTransform="uppercase"
```

Typography component support these props directly.

**Comment by koechkevin** [comment]
@kilemensi

File: apps/charterafrica/src/components/ErrorPage/ErrorPage.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+          variant="p3"
+          elements={description}
+        />
+        <RichTypography component="div" textAlign="center">
+          <Link underline="none" href={action.href || ""}>
+            <Button sx={{ mt: 5 }} color="secondary" variant="contained">
+              {action?.title}
+            </Button>
+          </Link>
+        </RichTypography>
```
Comment (kilemensi) [inline apps/charterafrica/src/components/ErrorPage/ErrorPage.js:66]: Just ```Button component={Link} sx={{ typography: xxxx }}``` should achieve what you want here.

File: apps/charterafrica/src/components/ErrorPage/ErrorPage.test.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+import ErrorPage from "./ErrorPage";
+
+import theme from "@/charterafrica/theme";
+
+// eslint-disable-next-line testing-library/render-result-naming-convention
+const render = createRender({ theme });
+
+const defaultProps = {
+  title: "Explainers",
```
Comment (kilemensi) [inline apps/charterafrica/src/components/ErrorPage/ErrorPage.test.js:12]: ?

File: apps/charterafrica/src/components/Layout/Layout.js [hunk 10]
```diff
@@ -8,6 +8,7 @@ import NavBar from "@/charterafrica/components/NavBar";
 
 const Main = styled("main")({
   flex: 1,
+  position: "relative",
```
Comment (kilemensi) [inline apps/charterafrica/src/components/Layout/Layout.js:11]: Still not sure why we need to switch this.

File: apps/charterafrica/src/lib/data/fetchApi.js [hunk 11]
```diff
@@ -0,0 +1,68 @@
+import qs from "qs";
```
Comment (kilemensi) [inline apps/charterafrica/src/lib/data/fetchApi.js:1]: Do we need any of this **_now_**?

File: apps/charterafrica/src/lib/data/fetchApi.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      this.headers = headers;
+      return;
+    }
+    this.authenticate();
+  }
+
+  async authenticate() {
+    const data = { email: this.email, password: this.password };
+    const res = await this.post(
+      `${process.env.NEXT_PUBLIC_APP_URL}/api/users/login`,
```
Comment (kilemensi) [inline apps/charterafrica/src/lib/data/fetchApi.js:25]: While they'll point to the same thing, any call meant to go to the Payload CMS should use `PAYLOAD_PUBLIC_APP_URL`.

File: apps/charterafrica/src/lib/data/fetchData.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+export const fetchSettings = async (args) => {
+  return fetchGlobals("settings", args);
+};
+
+export const fetchErrorPages = async (args) => {
+  return fetchCollection("errors", args);
+};
+
+export const fetchGlobalProps = async (args) => {
+  const settings = await fetchSettings(args);
```
Comment (kilemensi) [inline apps/charterafrica/src/lib/data/fetchData.js:54]:
```suggestion
  const settings = await fetchGlobals("settings, args);
```

I don't think we need any of the fetchXXXSpecificXXX. The above is simple enough for globals.

File: apps/charterafrica/src/pages/404.page.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  const { footer, navbar } = await fetchGlobalProps(args);
+
+  const { docs } = await fetchPage("404");
+  const [{ blocks: pages, slug }] = docs;
+  const setData = pages.find(({ statusCode }) => statusCode === 404);
+  const notFoundBlock = {
+    slug,
+    action: {},
+    ...setData,
+  };
```
Comment (kilemensi) [inline apps/charterafrica/src/pages/404.page.js:29]: 1. Check if page exist, through an error if it doesn't
2. Fetch globals
3. Profit.

File: apps/charterafrica/src/payload/blocks/Errors.js [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+      doc: {
+        ...siblingData.doc,
+        relationTo: "pages",
+        breadcrumbs: siblingData.doc.breadcrumbs || [],
+      },
+      linkType: "internal",
+    },
+    req,
+  });
+};
```
Comment (kilemensi) [inline apps/charterafrica/src/payload/blocks/Errors.js:17]:
```suggestion
```

You don't need this.

**Review by kilemensi — COMMENTED** [review]
👍🏽

File: apps/charterafrica/.env.template (hunk 2 above)
Comment (koechkevin) [reply apps/charterafrica/.env.template:24]: On the API, I get this error without any authentication

image

Comment (kilemensi) [reply apps/charterafrica/.env.template:24]: I ran the PR locally without providing my email/password and it worked. May be it was that hard-coded json file. Let me try again @koechkevin

Comment (kilemensi) [reply apps/charterafrica/.env.template:24]: Just tried again and I can pull the content without auth info @koechkevin

File: apps/charterafrica/src/lib/payload/staticData.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
+const notFoundData = {
```
[the remark this answers is not in the document]
Comment (koechkevin) [reply apps/charterafrica/src/lib/payload/staticData.js]: Yes we do not.

File: apps/charterafrica/src/components/Layout/Layout.js (hunk 10 above)
Comment (koechkevin) [reply apps/charterafrica/src/components/Layout/Layout.js:11]: Because I need to use absolute positioning somewhere in the child.

Comment (kilemensi) [reply apps/charterafrica/src/components/Layout/Layout.js:11]: ... that's kind of the question @koechkevin . We only need to set one background image, do we really need absolute positioning? Isn't `background-position` enough?

File: apps/charterafrica/src/lib/data/fetchApi.js (hunk 11 above)
Comment (koechkevin) [reply apps/charterafrica/src/lib/data/fetchApi.js:1]: We need a util function that converts nested objects searchParams.

`qs` does that.

Or we have to write a util function to handle that.

Comment (koechkevin) [reply apps/charterafrica/src/lib/data/fetchApi.js:1]: @kilemensi

Comment (kilemensi) [reply apps/charterafrica/src/lib/data/fetchApi.js:1]: Not sure I follow @koechkevin: How do nested search params look like in a URL?

Comment (koechkevin) [reply apps/charterafrica/src/lib/data/fetchApi.js:1]: We need to send a query with params object that looks like this 
```js
{
  locale: 'en',
  'fallback-locale': 'en',
  locales: [ 'en', 'fr', 'pt' ],
  where: { slug: { equals: '404' } }
}
```
We then expect a url like `?locale=en&fallback-locale=en&locales%5B0%5D=en&locales%5B1%5D=fr&locales%5B2%5D=pt&where%5Bslug%5D%5Bequals%5D=404`

but URLSearchParams ends up with something like `'?locale=en&fallback-locale=en&locales=en%2Cfr%2Cpt&where=%5Bobject+Object%5D'`

Comment (kilemensi) [reply apps/charterafrica/src/lib/data/fetchApi.js:1]: Damn, that's more like GraphQL-ing search params @koechkevin! `...&slug=404` should have been enough for REST API.

... but if that's the structure Payload expects, there is nothing we can do now.

🚀

File: apps/charterafrica/.env.template (hunk 2 above)
Comment (koechkevin) [reply apps/charterafrica/.env.template:24]: Could it be something I am missing?  I keep getting this.
image

Comment (koechkevin) [reply apps/charterafrica/.env.template:24]: similar t when I try to access https://charterafrica.prod.codeforafrica.org/api/globals/footer

Unless I register and login.

Comment (kilemensi) [reply apps/charterafrica/.env.template:24]: Could be... Why are you connecting to prod though? Have you tried dev or local/Docker payload @koechkevin ?

Comment (kilemensi) [reply apps/charterafrica/.env.template:24]: Does it work on local/Docker @koechkevin ?

Comment (koechkevin) [reply apps/charterafrica/.env.template:24]: Not connecting to prod.

Just that the responses are similar.

Comment (koechkevin) [reply apps/charterafrica/.env.template:24]: Ah!

Just found out the problem

File: apps/charterafrica/src/payload/blocks/Errors.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
+import linkGroup from "../fields/linkGroup";
+import richText from "../fields/richText";
+
+const Errors = {
+  slug: "errors",
```
Comment (kilemensi) [inline apps/charterafrica/src/payload/blocks/Errors.js:5]:
```suggestion
const Error = {
  slug: "error",{
    singular: {
      en: "Error",
      fr: "Erreur",
      pt: "Erro",
    },
    plural: {
      en: "Errors",
      fr: "Les Erreurs",
      pt: "Erros",
    },
  },
```

**Review by kilemensi — COMMENTED** [review]
👍🏽 @koechkevin 

-----

Good stuff... Almost there.

Restructure the `lib/data` into `lib/data/local` and `lib/data/rest`. As their names imply, they'll use their namesake APIs to fetch data from Payload. Put all common code i.e. code that process data after fetching, into `lib/data/common`. This way, the only thing different the two libs will be how data is fetched.


`lib/data/local` should export `getPageServerSideProps` while `lib/data/rest` can export `getPageStaticProps`. `lib/data/index.js` will then export both `getPageServerSideProps` for current pages to continue using without breaking, and `getPageStaticProps` to be used by error pages since they need to be created at build time (and any future page that needs to be static by design or performance).

File: apps/charterafrica/src/pages/_error.page.js [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+      case "500":
+        return <Error key={block.slug} {...block} />;
+      default:
+        return null;
+    }
+  });
+}
+// TODO find a way to deal with other status codes
+// Status code is only accessible on getStaticProps but does not work with payload
+export async function getStaticProps(args) {
```
[the remark this answers is not in the document]
Comment (koechkevin) [reply apps/charterafrica/src/pages/_error.page.js:19]: `getInitialProps` is missing crucial arguments such as `locale`  @kilemensi

File: apps/charterafrica/src/pages/_document.page.js [hunk 19]
```diff
@@ -114,7 +114,7 @@ MyDocument.getInitialProps = async (ctx) => {
     originalRenderPage({
       enhanceApp: (App) =>
         function EnhanceApp(props) {
-          return <App emotionCache={cache} {...props} />;
+          return <App emotionCache={cache} {...props} res={ctx.res} />;
```
[the remark this answers is not in the document]
Comment (koechkevin) [reply apps/charterafrica/src/pages/_document.page.js:117]: I should remove this.

File: apps/charterafrica/src/pages/_error.page.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (kilemensi) [reply apps/charterafrica/src/pages/_error.page.js:19]: What I'm trying to bring to your attention @koechkevin is the fact that `_error` needs to be rendered whenever an error occurs; client-side, server-side, at runtime. This means `getStaticProps` **should not** be used i.e. if we've set `statusCode` to **500** and generate html file for that at build time (that's what `getStaticProps` does), what will happen when **501** error occurs?

As long as you've read the [caveats](https://nextjs.org/docs/advanced-features/custom-error-page#caveats), if `getServerSideProps` work, great lets use that. Otherwise, we'll have to use `getInitialProps`.... with a bit of planning ahead, we can forward `locale` information to `getInitialProps` through the [custom](https://github.com/vercel/next.js/discussions/20242#discussioncomment-232060) [app](https://stackoverflow.com/a/65438136).

**Comment by kilemensi** [comment]
@kelvinkipruto please review.

**Review by kelvinkipruto — COMMENTED** [review]
Awesome stuff @koechkevin LGTM

**Review by kelvinkipruto — APPROVED** [review]

**Review by kilemensi — APPROVED** [review]
🚀

**Comment by kilemensi** [comment]
Lets get this merge @koechkevin
